subsys/usb/device_next/class/bt_hci.c

```diff
@@ -0,0 +1,491 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ * Copyright (c) 2023 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/* Bluetooth HCI USB transport layer implementation with following
+ * endpoint configuration:
+ *  - HCI commands through control endpoint (host-to-device only)
+ *  - HCI events through interrupt IN endpoint
+ *  - ACL data through one bulk IN and one bulk OUT endpoints
+ *
+ * TODO: revise transfer handling so that Bluetooth net_bufs can be used
+ * directly without intermediate copying.
+ *
+ * Limitations:
+ *  - Remote wakeup before IN transfer is not yet supported.
+ *  - H4 transport layer is not yet supported
+ */
+
+#include <zephyr/init.h>
+#include <zephyr/sys/byteorder.h>
+
+#include <zephyr/usb/usbd.h>
+#include <zephyr/usb/usb_ch9.h>
+#include <zephyr/drivers/usb/udc.h>
+
+#include <zephyr/net/buf.h>
+
+#include <zephyr/bluetooth/buf.h>
+#include <zephyr/bluetooth/hci_raw.h>
+#include <zephyr/bluetooth/l2cap.h>
+#include <zephyr/bluetooth/hci_vs.h>
+#include <zephyr/drivers/bluetooth/hci_driver.h>
+#include <zephyr/sys/atomic.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(bt_hci, CONFIG_USBD_BT_HCI_LOG_LEVEL);
+
+#define USB_BT_HCI_SUBCLASS		0x01
+#define USB_BT_HCI_PROTOCOL		0x01
+
+#define BT_HCI_DEFAULT_BULK_EP_MPS	0	/* Get maximum supported */
```

Comment:
I don't understand the comment - Can you elaborate?

Reply:
The Max Packet Size for bulk endpoints is speed-dependent and device only knows the operating speed after it is connected to host. There is no problem for Full-Speed devices, which always operate as Full-Speed and thus have 64 bytes Max Packet Size for bulk endpoints. However High-Speed devices can operate as both Full-Speed or High-Speed devices depending on connected host. When the stack is parsing endpoint descriptor and sees 0 as max packet size for bulk endpoint, it assigns either 64 (Full-Speed) or 512 (High-Speed).

Reply:
Ah, so when the MPS = `BT_HCI_DEFAULT_BULK_EP_MPS`, then it indicates that we should assigned the maximum supported value?

Reply:
"assigned" by the next USB device stack, in the sense that the device stack replaces the value with the max possible at connected speed.